Overlays

DalwynWasTaken opened the pull request:
Firstly I want to apologise for the previous PR being closed, I mis selected an option in my Git GUI and I by accidentally pushed a reverting commit to my branch which caused no changes to be seen which lead to the closing of the PR.

Now if you don't know what happened in PR 105:

Added OverlayPosition (To store overlay location and [NEW] if the overlay is toggled or not (to be used by ConfigAPI))
Added Overlay annotation (To be used at an OverlayPositon field to register an Overlay)
Added OverlayHolder to store Overlays (now integrated with ConfigManager to allow cross session saves)

Added OverlayManager to process these changes
Changed FeatureRegistry to reflect the new changes

Removed old Overlay class and moved it to overlays package.

-----

Again @P0keDev I want you to be aware of certain things within this change and that being getToggled and setToggled, this will allow ConfigAPI to make changes to the value that the feature has access to so they can toggle it from within their code. This forced me to add description, subcategory and visible to the overlay annotation since the toggle is something the user should see. I changed the serializer to match this. So you can handle Overlays as you would other ConfigOptions with a few extra steps to get the toggled state.

For those who noticed, it now looks very similar to Configs but merging would be a bad idea and cluster code that should be separated. 

And Magicus I attempted to keep Overlay as general as possible so you can make infolines from it but I believe Info lines should be done specially and simply have 1 OverlayPosition it exposes to the feature which is the container that stores the lines so it can move the container around.

I hope this provides a framework that meets everyone's needs for Overlays, if you want I can provide an example feature that has an Overlay or two in it.

P0keDev commented on common/src/main/java/com/wynntils/core/overlays/objects/OverlayPosition.java:
```diff
@@ -0,0 +1,137 @@
+                + "]";
+    }
+
+    public static class OverlayPositionSerializer
```
Don't forget to register this serializer in ConfigManager so it actually gets used

DalwynWasTaken committed:
Overlays

Firstly I want to apologise for the previous PR being closed, I mis selected an option in my Git GUI and I by accidentally pushed a reverting commit to my branch which caused no changes to be seen which lead to the closing of the PR.

Now if you don't know what happened in PR 105:

Added OverlayPosition (To store overlay location and [NEW] if the overlay is toggled or not (to be used by ConfigAPI))
Added Overlay annotation (To be used at an OverlayPositon field to register an Overlay)
Added OverlayHolder to store Overlays (now integrated with ConfigManager to allow cross session saves)

Added OverlayManager to process these changes
Changed FeatureRegistry to reflect the new changes

Removed old Overlay class and moved it to overlays package.

DalwynWasTaken committed:
Bug Fix

Change `field` to `parent.getClass()` to get the annotation needed.

P0keDev commented on common/src/main/java/com/wynntils/core/overlays/objects/OverlayHolder.java:
```diff
@@ -0,0 +1,39 @@
+        return metadata;
+    }
+
+    public boolean getToggled() {
```
I feel like these should be named "enabled" instead of "toggled" to match the language of the rest of the toggle-able aspects of the mod so far

DalwynWasTaken replied:
When making it I started realizing the it can be added to FeatureRegistry, and yeah I can change it all to enabled no problem. I'll work on it today.

P0keDev reviewed (dismissed):
Overall I think the code itself is fine, but I'm conflicted as to whether the `OverlayManager` is really necessary - it could probably be merged into `FeatureRegistry`, and the remaining files could be placed in the feature package. I'd like to hear the others' thoughts on it

magicus commented:
I still have a hard time seeing how this will be used in practice. I will repeat the same request that I made on the previous PR: Can you add an example implementation of an Overlay using this framework? I think it's close to impossible to judge the merits of this otherwise. The actual Overlay itself can be trivial. For all I care it can be just a dummy writing "HELLO WORLD" on the screen.

magicus commented on common/src/main/java/com/wynntils/core/config/objects/ConfigHolder.java:
```diff
@@ -16,7 +16,7 @@ public ConfigHolder(Object parent, Field field, Config metadata) {
                 parent,
                 field,
                 field.getType(),
-                field.getAnnotation(FeatureInfo.class).category(),
```
What's the purpose of this? Are you moving the annotation from the field to the class? 🤔

P0keDev replied:
the original line was incorrect - it was trying to get the `@FeatureInfo` annotation of the config field, which won't exist because the annotation is on the parent class

magicus commented on common/src/main/java/com/wynntils/core/features/FeatureRegistry.java:
```diff
@@ -104,6 +95,11 @@ private static void initializeFeature(Feature feature) {
         }
 
+        // register overlays
+        for (Field f : FieldUtils.getFieldsWithAnnotation(feature.getClass(), Overlay.class)) {
```
So are you supposed to declare Overlay fields in a Feature? As I said: we need an example on how you are planning to use this!

magicus commented on common/src/main/java/com/wynntils/core/features/FeatureRegistry.java:
```diff
@@ -184,94 +174,5 @@ public Object generate() {
-        }
-
-        @SubscribeEvent
-        public static void onRenderPre(RenderEvent.Pre e) {
```
Where did the rendering go? Or did you remove all this code? Why?

magicus commented on common/src/main/java/com/wynntils/core/overlays/objects/OverlayHolder.java:
```diff
@@ -0,0 +1,39 @@
+
+    public boolean getToggled() {
+        // Type casting to allow #isToggled to be called, it should be one anyway
+        return ((OverlayPosition) getValue()).isToggled();
```
I have so many questions about this. What is it that should be toggled? Visibility? And why not using standard bean conventions with `is` instead of `get` for booleans? And this casting looks really sus. If you are sure the value is of type OverlayPosition, why not assert it in the constructor? You're just passing on a `field.getType()`.

DalwynWasTaken replied:
I'm more than happy to replace all `get` to `is`

P0keDev commented:
> I still have a hard time seeing how this will be used in practice. I will repeat the same request that I made on the previous PR: Can you add an example implementation of an Overlay using this framework? I think it's close to impossible to judge the merits of this otherwise. The actual Overlay itself can be trivial. For all I care it can be just a dummy writing "HELLO WORLD" on the screen.

@DalwynWasTaken can correct me if I'm wrong, but the intended design for this framework is to declare one or more `OverlayPosition` variables in a feature, and then use whatever appropriate render events within the feature to draw the overlay, using the position defined in the corresponding position variable. Each `OverlayPosition` gets registered so that the user can customize where it is on the screen and whether it's enabled

magicus commented:
@P0keDev If that is indeed the purpose, I'm not sure I think that is a good idea. Any reasonable definition of an Overlay (if we even should have it as such) need to take care of its own rendering. Otherwise, what's the abstraction good for? Just storing screen locations? Seems like a case of the swedish proverb: "much screaming and almost no wool, said the farmer who tried to shear the pig".

magicus commented:
I'll repeat it once more, for effect. :-) Include an example of this being used, in the PR.

P0keDev commented:
> @P0keDev If that is indeed the purpose, I'm not sure I think that is a good idea. Any reasonable definition of an Overlay (if we even should have it as such) need to take care of its own rendering. Otherwise, what's the abstraction good for? Just storing screen locations? Seems like a case of the swedish proverb: "much screaming and almost no wool, said the farmer who tried to shear the pig".

At this point there's not much abstraction at all - just a special type of positional setting for overlay rendering. Maybe it would be best to reduce the framework even further: annotate the `OverlayPosition` like a normal config option, but collect them in a different list in `ConfigManager` so they can be easily accessed in the (eventual) HUD config screen. Essentially:

ConfigManager parses through every `@Config` field -> if field is of type `OverlayPosition`, place in `overlayOptions` list, otherwise place in `configOptions` list
Config GUI gets `configOptions` from ConfigManager, builds config GUI from them with category/metadata info
HUD editor gets `overlayOptions` from ConfigManager, builds HUD GUI from them with position info

DalwynWasTaken commented:
Hi, So I realize that this wasn't too clear. I forgot to add a usage example (it was in 105) to this PR which I should've done now taking into account what everyone has said: It can be now done in 2 ways. We add Overlays to FeatureRegistry or we add them to ConfigManager, both are reasonably simple to implement since they merely require the moving of some code.

```
package com.wynntils.features.user;

import com.wynntils.core.config.properties.Config;
import com.wynntils.core.features.UserFeature;
import com.wynntils.core.features.properties.EventListener;
import com.wynntils.core.features.properties.FeatureInfo;
import com.wynntils.core.features.properties.FeatureInfo.Stability;
import com.wynntils.core.config.objects.OverlayPosition;
import com.wynntils.mc.event.RenderEvent;
import com.wynntils.utils.objects.CustomColor;
import net.minecraftforge.eventbus.api.SubscribeEvent;


// Category and Visible come from Configurable and are used to place them in the config menu, visible defaults to true
@EventListener
@FeatureInfo(stability = Stability.STABLE, category = "This shouldn't be a raw string as brought up in issue 98")
public class TestFeature extends UserFeature {
    // Turning visible to true will reveal the internal boolean "enabled" in Overlay to the config api to show
    // not the raw x and y stuff, the reason for this is so that the user only needs 1 field to access all the data of
    // an overlay. If it's false, then you cannot "disable" the overlay as a user
    @Config(displayName = "Helo World Overlay", visible = true) // visible is default true
    // So x is 50% across the screen, y is 60% down the screen, width and height are just pixel values
    // For text I believe the default pixel size is 8-9 height and 2 width per character so this is just random
    // Enabled is true which means that the player can see this
    private final Overlay overlay = new Overlay(0.5,0.6, 100, 50, true);

    // This is just to show that it works with other config, you could also add another overlay to this feature without it
    // causing issues.
    @Config(displayName = "Hello World Text Color")
    private final CustomColor textColor = new CustomColor(255,255,255);
    
    @SubscribeEvent
    // Alright so you might've seen rendering stuff going poof, the reason being is that well it adds to value, the
    // code isn't shorter and it adds a middleman. Here you register to a render event of choice (if you are making
    // a certain feature you could then easily make your own) and then add the rendering code like before
    public void onRender(RenderEvent.Pre e) {
        //hmm I haven't looked into how you want to do rendering so I'll just use a fake function to prove my point
        if (!overlay.isEnabled()) return;

        // if this was real you wouldn't pass width and height since they are inferred with strings
        RenderUtils.drawString2D(
            "Hello World!",
            overlay.getCurrentX(),
            overlay.getCurrentY(), 
            overlay.getWidth(),
            overlay.getHeight());
    }
}
```
This is what it looks like in it's current state, if we moved it to featureregistry it would still be like this, if it was added to ConfigManager directly it would change the Overlay annotation to a Config annotation.

DalwynWasTaken commented:
> > @P0keDev If that is indeed the purpose, I'm not sure I think that is a good idea. Any reasonable definition of an Overlay (if we even should have it as such) need to take care of its own rendering. Otherwise, what's the abstraction good for? Just storing screen locations? Seems like a case of the swedish proverb: "much screaming and almost no wool, said the farmer who tried to shear the pig".
> 
> At this point there's not much abstraction at all - just a special type of positional setting for overlay rendering. Maybe it would be best to reduce the framework even further: annotate the `OverlayPosition` like a normal config option, but collect them in a different list in `ConfigManager` so they can be easily accessed in the (eventual) HUD config screen. Essentially:
> 
> ConfigManager parses through every `@Config` field -> if field is of type `OverlayPosition`, place in `overlayOptions` list, otherwise place in `configOptions` list Config GUI gets `configOptions` from ConfigManager, builds config GUI from them with category/metadata info HUD editor gets `overlayOptions` from ConfigManager, builds HUD GUI from them with position info

You'll need to decide if it gets added to FeatureRegistry (Like they were before) or ConfigManager. I think ConfigManager stays close to home since the entire reason for this annotation stuff is to allow for easier storage of overlays

DalwynWasTaken committed:
Merging with Config

Removed Overlay annotation
Made Config the main annotation and if it picks up an Overlay type as the field, it registers an OverlayHolder

Changed OverlayPosition to Overlay due to recent changes to the class and it storing everything you need for an overlay

Changed some comments for clearer, more descriptive comments.

Removed overlays package since it was merged with config package

Removed OverlayManage and Overlay annotation since they are now unnecessary.

DalwynWasTaken commented:
In regards to the new commit, there still needs to be a getOverlays method, it can be added to FeatureRegistry or ConfigManager

Option 1:
```
// FeatureRegistry file

// At the top of the class
private static final List<OverlayHolder> OVERLAYS = new ArrayList<>();

// At where config is registered
if (f.getType() == Overlay.class) {
    OverlayHolder overlay = new OverlayHolder(feature.getClass(), f, f.getAnnotation(Config.class));
    OVERLAYS.add(overlay)
    ConfigManager.registerHolder(overlay);
}

// New method at the bottom
public static List<OverlayHolder> getOverlays() {
    return OVERLAYS;
}
```

Option 2:

```
// ConfigManager file

// At the top of the class
private static final List<OverlayHolder> OVERLAYS = new ArrayList<>();

// registerHolder method
if (holder instanceof OverlayHolder) {
    OVERLAYS.add(holder);
    STORAGE_HOLDERS.add(holder);
}

// new method
public static List<OverlayHolder> getOverlays() {
     return OVERLAYS;
}
```

Option 3: (The one I would implement)
```
// ConfigManager file

// new method
public static List<OverlayHolder> getOverlays() {
     List<OverlayHolder> overlays = new ArrayList<>();
     for (StorageHolder holder : STORAGE_HOLDERS) {
         if (holder instanceof OverlayHolder) {
               overlays.add(holder);
         }
     }
     return overlays
}
```

DalwynWasTaken commented:
A logical next step for overlays would be adding rendering methods to RenderUtils to help drawing common things that you would do with an overlay (Rectangles are already a thing but why can you define Z? If that's to show on which level it should be drawn do you want me to add that to Overlay?) mainly strings.

I also like Magicus's idea of generalizing Info Lines, I feel like instead of each feature making their own overlay they could just "push" a message to info lines, this also removes more usage of random overlays and centralizes info messages.

DevScyu commented:
> In regards to the new commit, there still needs to be a getOverlays method, it can be added to FeatureRegistry or ConfigManager
> 
> Option 1:
> 
> ```
> // FeatureRegistry file
> 
> // At the top of the class
> private static final List<OverlayHolder> OVERLAYS = new ArrayList<>();
> 
> // At where config is registered
> if (f.getType() == Overlay.class) {
>     OverlayHolder overlay = new OverlayHolder(feature.getClass(), f, f.getAnnotation(Config.class));
>     OVERLAYS.add(overlay)
>     ConfigManager.registerHolder(overlay);
> }
> 
> // New method at the bottom
> public static List<OverlayHolder> getOverlays() {
>     return OVERLAYS;
> }
> ```
> 
> Option 2:
> 
> ```
> // ConfigManager file
> 
> // At the top of the class
> private static final List<OverlayHolder> OVERLAYS = new ArrayList<>();
> 
> // registerHolder method
> if (holder instanceof OverlayHolder) {
>     OVERLAYS.add(holder);
>     STORAGE_HOLDERS.add(holder);
> }
> 
> // new method
> public static List<OverlayHolder> getOverlays() {
>      return OVERLAYS;
> }
> ```
> 
> Option 3: (The one I would implement)
> 
> ```
> // ConfigManager file
> 
> // new method
> public static List<OverlayHolder> getOverlays() {
>      List<OverlayHolder> overlays = new ArrayList<>();
>      for (StorageHolder holder : STORAGE_HOLDERS) {
>          if (holder instanceof OverlayHolder) {
>                overlays.add(holder);
>          }
>      }
>      return overlays
> }
> ```

Option 2 seems the best in my opinion, option 3 would be looping over the storage holders every time getOverlays() was called, when we could just cache it on init.

kristofbolyai commented on common/src/main/java/com/wynntils/core/config/objects/ConfigHolder.java:
```diff
@@ -16,7 +16,7 @@ public ConfigHolder(Object parent, Field field, Config metadata) {
                 field,
                 field.getType(),
-                field.getAnnotation(FeatureInfo.class).category(),
+                parent.getClass().getAnnotation(FeatureInfo.class).category(),
```
This line assumes every feature has a @FeatureInfo, which is not enforced atm, so it crashes whenever it hits a feature like that. Fix that by defaulting the category to null if `parent.getClass().getAnnotation(FeatureInfo.class)` is `null`.

kristofbolyai replied:
Regarding this, do you think we should not allow loading features without a FeatureInfo annotation? This is not subject to this PR, but I think it might be helpful to do it in the long run.

P0keDev replied:
at the very least, we shouldn't load config options from features with no FeatureInfo annotation - it might be better to change the constructor to take a string for the category instead of determining it within the constructor. i plan on making a lot of changes/improvements wrt config loading soon, so this doesn't need to be dealt with in this PR

magicus replied:
It's probably not a secret that I don't think the FeatureInfo provides much value (at least not now -- I am not against introducing annotations for features if that ever changes in the future), and I am questioning their need. So I definitely do not want to see them enforced. :-)

DalwynWasTaken committed:
Small tweaks

Added a null check to the annotation getter for FeatureInfo

Added a separate storage for OverlayHolders in ConfigManager
Added a getter for OverlayHolders in ConfigManager

kristofbolyai commented:
`StorageHolder` crash with UserFeatures. Removing the config annotation from `userEnabled` "fixes" the crash.

```
java.lang.RuntimeException: Could not execute entrypoint stage 'client' due to errors, provided by 'wynntils'!
	at net.fabricmc.loader.impl.entrypoint.EntrypointUtils.lambda$invoke0$0(EntrypointUtils.java:51)
	at net.fabricmc.loader.impl.util.ExceptionUtil.gatherExceptions(ExceptionUtil.java:33)
	at net.fabricmc.loader.impl.entrypoint.EntrypointUtils.invoke0(EntrypointUtils.java:49)
	at net.fabricmc.loader.impl.entrypoint.EntrypointUtils.invoke(EntrypointUtils.java:35)
	at net.fabricmc.loader.impl.game.minecraft.Hooks.startClient(Hooks.java:53)
	at net.minecraft.client.Minecraft.<init>(Minecraft.java:452)
	at net.minecraft.client.main.Main.main(Main.java:197)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at net.fabricmc.loader.impl.game.minecraft.MinecraftGameProvider.launch(MinecraftGameProvider.java:416)
	at net.fabricmc.loader.impl.launch.knot.Knot.launch(Knot.java:77)
	at net.fabricmc.loader.impl.launch.knot.KnotClient.main(KnotClient.java:23)
	at net.fabricmc.devlaunchinjector.Main.main(Main.java:86)
	at dev.architectury.transformer.TransformerRuntime.main(TransformerRuntime.java:217)
Caused by: java.lang.IllegalArgumentException: Can not set boolean field com.wynntils.core.features.UserFeature.userEnabled to java.lang.Class
	at java.base/jdk.internal.reflect.UnsafeFieldAccessorImpl.throwSetIllegalArgumentException(UnsafeFieldAccessorImpl.java:167)
	at java.base/jdk.internal.reflect.UnsafeFieldAccessorImpl.throwSetIllegalArgumentException(UnsafeFieldAccessorImpl.java:171)
	at java.base/jdk.internal.reflect.UnsafeFieldAccessorImpl.ensureObj(UnsafeFieldAccessorImpl.java:58)
	at java.base/jdk.internal.reflect.UnsafeBooleanFieldAccessorImpl.getBoolean(UnsafeBooleanFieldAccessorImpl.java:40)
	at java.base/jdk.internal.reflect.UnsafeBooleanFieldAccessorImpl.get(UnsafeBooleanFieldAccessorImpl.java:36)
	at java.base/java.lang.reflect.Field.get(Field.java:425)
	at org.apache.commons.lang3.reflect.FieldUtils.readField(FieldUtils.java:428)
	at com.wynntils.core.config.objects.StorageHolder.getValue(StorageHolder.java:51)
	at com.wynntils.core.config.objects.StorageHolder.<init>(StorageHolder.java:30)
	at com.wynntils.core.config.objects.ConfigHolder.<init>(ConfigHolder.java:15)
	at com.wynntils.core.features.FeatureRegistry.initializeFeature(FeatureRegistry.java:97)
	at com.wynntils.core.features.FeatureRegistry.registerFeature(FeatureRegistry.java:51)
	at com.wynntils.core.features.FeatureRegistry.init(FeatureRegistry.java:133)
	at com.wynntils.core.WynntilsMod.init(WynntilsMod.java:53)
	at com.wynntils.fabric.WynntilsModFabric.onInitializeClient(WynntilsModFabric.java:21)
	at net.fabricmc.loader.impl.entrypoint.EntrypointUtils.invoke0(EntrypointUtils.java:47)
	... 13 more
```

magicus commented:
@DalwynWasTaken Of your examples, I think option 1 is the best. I still think "Overlay" has the potential to be a much more encompassing concept than just "a set of coordinates for the user to change", and as such, it is an integral part of the framework we want to support our Features with.

I also think the Overlays should be in charge of their rendering. But we can take that in a separate PR, and let this one focus on "how to describe overlay size and position".

magicus commented:
When I look through the entire code diff once more, I must say I strongly dislike the way this PR changes the architecture. The `Overlay` class did reside in `core.features.overlays`, which is exactly were it belongs -- overlay support is part of the feature framework. Also, the rendering framework has been removed. It was probably misplaced in `FeatureRegistry`, but it should be moved to a better place (possibly the `Overlay` class itself), not removed.

This PR is going in the opposite direction we want to go -- demoting Overlays from a first class citizen to a mere configuration of a rectangle. Sure, the layout directions of Overlays are important, but that's just a part of what it means to be an overlay.

The new config Overlay class also has a lot of "suprises". Like `public static Overlay fromString(String string)` -- what is that supposed to achieve?

DalwynWasTaken commented:
I actually agree, I do feel like this is going the wrong way. I don’t believe that there should be a rendering framework but that depends on how simple we want to make a Feature then again, you people have more experience so I’ll stick to better judgment and keep in. I was thinking of restarting from scratch. As for the Overlay surprise. That is a result of lots of small patches without a greater overview. It will remain however but maybe not in that class. It’s used to store it across sessions. I also realise that my commenting has been subpar and this has turned into a mess which wasn’t the meaning. I shall take a step back now that I know what we’re going for and just write down the entire process and share it with you guys before I implement it. I also think that I made some good and bad changes to Configuration in the process that I need to fix. I hope to get that done before Poke does his rework, and then apply Overlay once Poke has done them. I think that Info Lines should be generalised and will try to incorporate some sort of structure for it.

I do have some questions for you @magicus You say that there is so much more to overlays, you’ll need to be more consice

DalwynWasTaken commented:
And I miss clicked. Well I shall continue. You need to be more precise so others know what you mean and want. I cannot mind read and some clearer directions could help further down the road and could you also share what you wish to see implemented and possibly how?

magicus commented:
Some random thoughs about Overlay:
 * They can either be completely new (like minimap), or replace vanilla Minecraft HUD features (like health bar).
 * Location reasonably need some kind of "snapping" behavior, so you can position overlays with respect to a side of the screen. (We never did this good in Legacy).
 * Overlays that are snapped to the right side of the screen will likely need to have right-aligned text etc. This should tie in with the snapping, but possibly also be overridable.
 * Some overlays might be resizable (e.g. the ticker/message pane), others perhaps needs to be fixed size (like HUD overlays, perhaps minimap etc).
 * Overlays conceptually will belong to a Feature, but a Feature might hold multiple Overlays (this is still open for discussion, but this is my current thinking).
 * Overlays should be implemented as a stand-alone component within a Feature, typically as a inner class, or where you register a render method handle with the overlay.
 * The render method for the Overlay must help the Overlay render within the Overlay box. This means it is not likely to be a direct translation from a Render.Post/Pre event, but more likely that a dispatcher draws enabled Overlays by calling their render methods.
 * We need to have a good toolbox of RenderUtils stuff to help us efficiently and correctly render the kind of stuff we need in overlays.
 * We might want to have different "layouts", with a different set of Overlays enabled, and different placements, and a simple way to switch between them.
 * To help onboarding, we might start with a few predefined layouts, like a "minimap only", "beginner friendly" and "hurt me plenty" with all the goodies.
 * Overlays might need to be dynamically created, like the info lines. Or perhaps we should not assume that there is only a single instance of each overlay -- while having two minimaps might sound odd, having a design that will allow that is bringing us flexibility that can come in handy for other situations.

magicus commented:
That was just a quick braindump. We can't implement all of that at once. And we should not have hubris to think that we can get it right the first time. But neither should we paint ourselves in a corner, and make such functionality hard or impossible to implement. 

And in this scheme of things, how to integrate the Overlay positioning information with other configuration is just a small piece of the puzzle.

DalwynWasTaken commented:
This is beautiful. I love some of the ideas you threw in there! That was an eye opener, I think tunnel vision started kicking in. Thanks for that

DalwynWasTaken commented:
I'm closing this as this isn't going in the direction that was intended and doesn't meet certain quality standards. #110 is to discuss overlays and their step by step buildup to something we all agree on.

I made this PR in hope to get some energy and thought into creating an overlay framework for this project, which it did do. Thank you for your input into this and putting up with me. I learned a lot from this. Thank you
